### ISSUE TITLE
Potential connection leak

### ISSUE THREAD

**Issue by minborg** [speedment/jpa-streamer#147]
A customer reports connections being leaked with spring boot and postgres, hiraki

**Comment by minborg** [comment]
Customer reports: 

" I solved the issue by manual open and close jpastream every query. Connection leak don't happening and everything on server is good working . But I hope you update version for fix this issue (connection leak)."

**Comment by sharmakrravi** [comment]
Yes, I also tried the latest 1.1.0 and still the connection leak is there with spring boot, hikari

**Comment by julgus** [comment]
Considering these reports are old, they may or may not be relevant. But any cues on how to reproduce this would be helpful. For now, I have tried running the jpastreamer-demo Spring Boot application with Hakari. 100,000 consecutive requests all succeed without any issues. I am not closing JPAStreamer in between calls. 

What symptoms to you see that indicates that there is a connection leak?

**Comment by julgus** [comment]
Okay, found the crack. The single connection leak that springs from a regular JPAStreamer Stream does not block the application and only shows up once in the beginning. Interestingly no connection leak appears whenever I only perform joined queries, thus I assume the issue is associated with the JPAStreamer caching of simple StreamConfiguration (no joins or projections). 

In contrast, triggering a reset of JPAStreamer on every request (`jpaStreamer.resetStreamer(Film.class)`) results in every connection being leaked, and the application becomes blocked after 21 requests (20 maximum connections). 

I'll keep digging.

**Comment by julgus** [comment]
Disabling the JPAStreamer internal Stream cache resolves the issue. Would however be nice to get to the bottom of why that cache is problematic.

**Comment by julgus** [comment]
My latest comment is not entirely wrong. But returning to this issue with fresh eyes helped me see the underlying problem more clearly. 

Hikari connections are coupled with the lifecycle of an Entity Manager. From the perspective of JPAStreamer, Hikari will not release a connection until we close the Entity Manager we have obtained. 

In the case where an application is repeatedly calling e.g. `jpaStreamer.stream(Film.class)` JPAStreamer caches the underlying `Streamer` associated with the Film entity to save resources. The `Streamer` has a reference to a `Renderer`, that in turn holds a reference to an `EntityManager`. As we are reusing the same Streamer over and over again, we are also reusing the same Entity Manager, and it will stay active until we close down JPAStreamer altogether (or resets the cache, see final paragraph). This also means that Hikari will reserve a connection for each Streamer in the cache and this will create leaks if a user creates a bunch of Streamers that are then never reused. Or is recognized as a leak in Hikari even if the connection is active, as it is never returned to the pool. 

Uncached Streamers associated with more complex Stream Configurations (joins, projections etc.) will close the underlying EntityManager upon completion as the Streamer is discareded. Therefore these are leakproof. 

The way I see this, it is often desired to reuse an EntityManager repeatedly to save resources, thus the cache is motivated. However, we need to offer the possibility to manually close the underlying EntityManager to release the connection, alternatively, disable the cache altogether. 

A while ago I introduced the option of manually removing a _Streamer_ instance from the cache by calling jpaStreamer.resetStreamer(Class<?>... entityClasses). What I overlooked at that point was that the underlying Entity Manager is never closed. Thus the current implementation of this method does not offer a solution to the problem but rather leaks even more connections. This will be fixed in my next commit.
